### PR TITLE
[1.12] Added AdvancementGrantEvent

### PR DESCRIPTION
**Pull request by MiraWaNeko**
This adds two events related to Advancements.

AdvancementEvent.Pre which triggers when a player has completed an advancement, but before the player receives the rewards. This is cancelable.
AdvancementEvent.Post which triggers after the player has received the rewards for completing an advancement.

### CODE REVIEW

**Comment by liach** [comment]
Rename it to GrantAdvancementEvent is better.

File: patches/minecraft/net/minecraft/advancements/PlayerAdvancements.java.patch [hunk 1]
```diff
@@ -0,0 +1,33 @@
+--- ../src-base/minecraft/net/minecraft/advancements/PlayerAdvancements.java
++++ ../src-work/minecraft/net/minecraft/advancements/PlayerAdvancements.java
+@@ -27,6 +27,8 @@
+ import net.minecraft.util.JsonUtils;
+ import net.minecraft.util.ResourceLocation;
+ import net.minecraft.util.text.TextComponentTranslation;
++import net.minecraftforge.common.MinecraftForge;
++import net.minecraftforge.event.entity.player.AdvancementEvent;
+ import org.apache.logging.log4j.LogManager;
```
Comment (Madsthunder) [inline patches/minecraft/net/minecraft/advancements/PlayerAdvancements.java.patch]: Delete these two imports...

File: patches/minecraft/net/minecraft/advancements/PlayerAdvancements.java.patch [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ import org.apache.logging.log4j.Logger;
+ 
+@@ -208,12 +210,21 @@
+ 
+             if (!flag1 && advancementprogress.func_192105_a())
+             {
++                if (MinecraftForge.EVENT_BUS.post(new AdvancementEvent.Pre(this.field_192762_j, p_192750_1_))) {
```
Comment (Madsthunder) [inline patches/minecraft/net/minecraft/advancements/PlayerAdvancements.java.patch]: ...change this line to `net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.AdvancementEvent.Pre(this.field_192762_j, p_192750_1_))) {`, ...

File: patches/minecraft/net/minecraft/advancements/PlayerAdvancements.java.patch [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
++                }
++
+                 p_192750_1_.func_192072_d().func_192113_a(this.field_192762_j);
+ 
+                 if (p_192750_1_.func_192068_c() != null && p_192750_1_.func_192068_c().func_193220_i() && this.field_192762_j.field_70170_p.func_82736_K().func_82766_b("announceAdvancements"))
+                 {
+                     this.field_192756_d.func_184103_al().func_148539_a(new TextComponentTranslation("chat.type.advancement." + p_192750_1_.func_192068_c().func_192291_d().func_192307_a(), new Object[] {this.field_192762_j.func_145748_c_(), p_192750_1_.func_193123_j()}));
+                 }
++
++                MinecraftForge.EVENT_BUS.post(new AdvancementEvent.Post(this.field_192762_j, p_192750_1_));
```
Comment (Madsthunder) [inline patches/minecraft/net/minecraft/advancements/PlayerAdvancements.java.patch]: ...And change this line to `net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.AdvancementEvent.Post(this.field_192762_j, p_192750_1_));`

**Review by Madsthunder — CHANGES_REQUESTED** [review]
The pr looks good, but it looks like some unnecessary patches were accidentally added.

**Comment by MiraWaNeko** [comment]
@liach I agree, I will commit the namechange in a moment
@Madsthunder It seems like you're suggesting changes that exist in the second commit. Can you check?

**Comment by MiraWaNeko** [comment]
@liach Upon further thinking, would `AdvancementGrantEvent` maybe be better?
This would make it more in line with other events where they starts with what the event overall is about and then the action, such as `LivingDeathEvent` and `LivingDropsEvent`.

**Comment by Madsthunder** [comment]
@Chikachi You're right, I had a slight derp...

File: src/main/java/net/minecraftforge/event/entity/player/AdvancementGrantEvent.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        this.advancement = advancement;
+    }
+
+    public Advancement getAdvancement()
+    {
+        return advancement;
+    }
+
+    @Cancelable
+    public static class Pre extends AdvancementGrantEvent
```
Comment (howtonotwin) [inline src/main/java/net/minecraftforge/event/entity/player/AdvancementGrantEvent.java:45]: There should be a Javadoc on this that explicitly states that this event is cancelable and what happens when this event is canceled.

**Review by howtonotwin — CHANGES_REQUESTED** [review]

**Comment by TechnicianLP** [comment]
- 1.13 will probably cause problems with this pr as advancements a mod expects can be missing/renamed via datapacks - nevertheless the `Post`-Event is usefull for other datapack-defined systems depending on advancements
- Why exactly is the Pre one neccessary? if you dont want the player to get the advancement yet  you should add an additional criterion to it instead

**Comment by MiraWaNeko** [comment]
The `Pre` isn't neccessary, I added it simply because there might be a use case that I hadn't thought about.

**Comment by BillyGalbreath** [comment]
Any word on this? I'm eager to use this API

**Comment by liach** [comment]
This patch needs to get fixed.

**Comment by mezz** [comment]
Superseded by #4467

**Pull request closed without merging**